Case tiles for case detail

orangejenny opened the pull request:
## Product Description
https://dimagi-dev.atlassian.net/browse/USH-3873

The related formplayer PR has already been deployed: https://github.com/dimagi/formplayer/pull/1530

Configuration in app manager:

<img width="910" alt="Screen Shot 2024-01-18 at 10 05 49 AM" src="https://github.com/dimagi/commcare-hq/assets/1486591/60a4fb92-9474-466a-b68c-492186170b3e">

Web Apps UI corresponding to that configuration:

<img width="614" alt="Screen Shot 2024-01-18 at 10 06 43 AM" src="https://github.com/dimagi/commcare-hq/assets/1486591/4060b5ab-8a45-43b3-844f-15fe638ad10a">
<img width="621" alt="Screen Shot 2024-01-18 at 10 06 50 AM" src="https://github.com/dimagi/commcare-hq/assets/1486591/72bdf7a3-40c1-4b4f-b60c-3c5d03687bb5">
<img width="621" alt="Screen Shot 2024-01-18 at 10 06 56 AM" src="https://github.com/dimagi/commcare-hq/assets/1486591/9f4f16dd-235b-482d-b471-f0be57940c3f">

## Feature Flag
New flag: USH: Case detail tile

## Safety Assurance

### Safety story
New feature extending an existing flagged feature. Risk is primarily to case tiles functionality.

### Automated test coverage

Tests in PR

### QA Plan

https://dimagi-dev.atlassian.net/browse/QA-6030
### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

orangejenny committed:
Added case tile config UI to case detail tab

orangejenny committed:
Updated module saving to save template for either short or long detail

orangejenny committed:
Removed validation limiting tiles to short detail

orangejenny committed:
Added tile to web apps UI

orangejenny committed:
Removed redundant code

orangejenny committed:
Added translation

orangejenny committed:
Updated case tile mapping dropdown to not show for tabs

orangejenny committed:
Removed tab properties from case tile preview

orangejenny committed:
Fixed display of tab rows while using custom tile

orangejenny committed:
Added suite support for case tiles in nested details

orangejenny committed:
Moved short detail logic behind if statement now that CaseTileHelper supports long details

orangejenny committed:
Added shading for tab regions of case tile preview

orangejenny committed:
Limited detail tiles to manual configuration

orangejenny committed:
Added display of case tiles for nodeset tabs

orangejenny committed:
Extracted initialization logic for views that list case tiles

orangejenny committed:
Updated tiles to be non-clickable when displayed in details

orangejenny committed:
Added validation that each tile row uses only one tab

orangejenny committed:
Added basic tests

MartinRiese commented on corehq/apps/app_manager/static/app_manager/js/details/screen.js:
```diff
@@ -150,6 +153,21 @@ hqDefine("app_manager/js/details/screen", function () {
+        // or an even tab, for the sake of being able to differentiate in the case tile preview
+        // which rows go with which tab.
+        self.tabPolarity = function (column) {
+            var self = this,
```
feel free to use `const` ;)

orangejenny replied:
thanks, I haven't developed good ES6 habits yet - 904070a6db4

orangejenny committed:
Fixed lint errors

MartinRiese commented on corehq/apps/app_manager/static/app_manager/js/details/screen.js:
```diff
@@ -150,6 +153,21 @@ hqDefine("app_manager/js/details/screen", function () {
+        self.tabPolarity = function (column) {
+            var self = this,
+                flag = false;
+            _.find(self.columns(), function (c, i) {
```
I could not verify that `_.find` is guarantied to run in over the items in sequence but since JS is single threaded I think this should be the case

orangejenny replied:
Interesting, I hadn't thought about that. I suppose `each` would guarantee the order, though it also wouldn't end early.

orangejenny replied:
Although, looking at the [docs](https://underscorejs.org/#find), I think "returning the first one that passes a truth test" implies that the items are examined in order.

MartinRiese commented on corehq/apps/app_manager/suite_xml/sections/details.py:
```diff
@@ -116,31 +116,23 @@ def get_section_elements(self):
-
-                            d = helper.build_case_tile_detail()
-                            self._add_custom_variables(detail, d)
+                        print_template_path = None
+                        if detail.print_template:
+                            print_template_path = detail.print_template['path']
```
good candidate to extract into a function to keep complexity down.

orangejenny replied:
extracted into a function, which was so tiny I decided to just inline it: 701e5cc2361

MartinRiese commented on corehq/toggles/__init__.py:
```diff
@@ -709,14 +709,24 @@ def _ensure_valid_randomness(randomness):
     parent_toggles=[CASE_LIST_TILE],
 )
 
+CASE_DETAIL_TILE = StaticToggle(
```
could that not be part of CASE_LIST_TILE_CUSTOM? If only USH uses it I don't really see the point in adding a new FF.

orangejenny replied:
700b0afd4ea

MartinRiese commented on corehq/apps/hqwebapp/static/app_manager/less/case_tile_preview.less:
```diff
@@ -13,6 +13,10 @@
         transition: 2s;
 
+        &.odd-tab {
+            background-color: darken(@cc-bg, 3);
```
Can we reuse the colors that Jonathan used for the groups?

orangejenny replied:
These ones? https://github.com/dimagi/commcare-hq/pull/33801

I like the thought of reusing colors, but those are so dark they seem a bit overwhelming:
<img width="1004" alt="Screen Shot 2024-01-18 at 2 10 59 PM" src="https://github.com/dimagi/commcare-hq/assets/1486591/20c2bd70-745b-48fa-b54f-68c5552d3050">

MartinRiese replied:
Oh. Agreed.

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/controller.js:
```diff
@@ -240,8 +246,12 @@ hqDefine("cloudcare/js/formplayer/menus/controller", function () {
                 title: detailObject.get('title'),
             };
+            if (detailObject.get('usesCaseTiles')) {
+                return views.CaseTileDetailView(menuData);
```
I think `CaseDetailTileView` would be more descriptive if we could also rename `CaseListDetailView` to `CaseDetailTableView`

orangejenny replied:
Tweaked colors ever so slightly to be more standard: 39cf74a9778

MartinRiese reviewed (commented):
Still have to look through `formplayer/menus/views.js`. will do that later.

orangejenny committed:
Updated test to use modern app version

orangejenny committed:
Updated tests to use AppFactory

Doing this because it encourages making the app with a recent CommCare version  - new_app deaults
to a very old version which triggers this code:
https://github.com/dimagi/commcare-hq/blob/d724c1cec867f3a599600e94b67ecc6d3d4f5a80/corehq/apps/app_manager/suite_xml/sections/details.py#L256

orangejenny committed:
Updated js

orangejenny committed:
Inlined print template logic

orangejenny committed:
Moved new functionality under CASE_LIST_TILE_CUSTOM flag

orangejenny committed:
Fixed JavaScript declarations

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -473,6 +473,34 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
+        rowClick: function () {},
+    });
+
+    var initCaseTileList = function (options) {
```
Nice. Could probably also be a `const` since the views are.

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -473,6 +473,34 @@ hqDefine("cloudcare/js/formplayer/menus/views", function () {
         },
     });
 
+    const CaseTileViewUnclickable = CaseTileView.extend({
```
Seems like the inheritance should be the other way around since `CaseTileView` has more functionality but I that would probably be tricky.

orangejenny replied:
I think this makes sense, so I played around with it, but now I'm inclined to leave it as-is. I did a commit that made `CaseViewUnclickable` the parent view of `CaseView`, which is fine. `CaseTileViewUnclickable` can then be a child of `CaseViewUnclickable`, but `CaseTileView` needs functionality from both `CaseView` and `CaseTileViewUnclickable` which gets messy.

MartinRiese committed:
Semicolon not comma

orangejenny committed:
Fixed js syntax

orangejenny committed:
Updated colors

orangejenny committed:
Merge branch 'jls/case-tile-for-case-detail' of github.com:dimagi/commcare-hq into jls/case-tile-for-case-detail

orangejenny committed:
Updated declarations

orangejenny committed:
Merge branch 'master' into jls/case-tile-for-case-detail

orangejenny committed:
Updated persistent tiles to be unclickable to screen readers

orangejenny committed:
Revert "Updated persistent tiles to be unclickable to screen readers"
